### ISSUE TITLE
Get rid of guava

### ISSUE THREAD

**Issue by XakepSDK** [auth0/jwks-rsa-java#120]
### Describe the problem you'd like to have solved
I want to get rid of guava. It's very big and not modularized.
Project i work on is 100% modularized and i need this library, but it uses guava.

### Describe the ideal solution
1. Get rid of Guava where possible
2. On modules side use `requires static` and in runtime check once if guava is available and if it's not available and guava cached provider requested, throw an exception, both features can be implemented using multi version jars

## Alternatives and current work-arounds
From this project side:
1. Get rid of Guava where possible, it's every class except `GuavaCachedJwkProvider`

From my project side:
1. Use build plugins to generate module descriptor
2. Strictly check to not use cached provider <- not sure how well this will work with modules

### Additional information, if any
I will create PR to get rid of guava where possible

**Comment by jimmyjames** [comment]
Thanks for the info and the PR. We are tracking this to review the PR internally with SDK-2551.

**Comment by lbalmaceda** [comment]
@XakepSDK I've merged that PR.

If you want your own implementation of a cached provider, you could have a class implement the `JwkProvider` interface and wrap another instance of `JwkProvider` following the composite pattern that we follow in our implementation (see the constructors). Of course, you won't be able to use the helper Jwk builder with it, as it makes use of Guava's implementation... 

Then, if you need to completely remove the Guava dependency from your build, provided you are not using the "Guava Provider" we include, you could use Gradle to avoid pulling that dependency. Something like this:

```groovy
compile('implementation 'com.auth0:jwks-rsa:0.17.1) {
    exclude group: 'com.google.guava', module: 'guava' 
}
```

Does that help?

**Comment by XakepSDK** [comment]
Yes, it should work. I also will add module descriptor using ModiTect maven plugin, it let's you add descriptor for ordinary maven dependency.
If this library will migrate to java 9+, caffeine could be used. It's modularized already.

**Comment by lbalmaceda** [comment]
@XakepSDK Thanks. I'll make a release with the change we merged past Friday and update this ticket. 

Unfortunately, none of our libraries are moving to Java 9 anytime soon. Java 8 is still widely used and we need to keep supporting it.

**Comment by lbalmaceda** [comment]
Released under `0.18.0`. Thanks

**Comment by Dmole** [comment]
@lbalmaceda can this be re-visited?
